worldedit-bukkit/src/main/java/com/fastasyncworldedit/bukkit/adapter/NMSAdapter.java

```diff
@@ -140,4 +144,122 @@ public void sendChunk(IChunkGet chunk, int mask, boolean lighting) {
         ((BukkitGetBlocks) chunk).send();
     }
 
+    private static final ConcurrentHashMap<String, ConcurrentHashMap<IntPair, ChunkLock>> SENDING_CHUNKS = new ConcurrentHashMap<>();
+
+    /**
+     * Atomically set the given chunk section to the chunk section array stored in the chunk, given the expected existing chunk
+     * section instance at the given layer position.
+     * <p>
+     * Acquires a (FAWE-implemented only) write-lock on the chunk packet lock, waiting if required before writing, then freeing
+     * the lock. Also sets a boolean to indicate a write is waiting and therefore reads should not occur.
+     * <p>
+     * Utilises ConcurrentHashMap#compute for easy synchronisation for all of the above. Only tryWriteLock is used in blocks
+     * synchronised using ConcurrentHashMap methods.
+     *
+     * @since TODO
+     */
+    protected static <LevelChunkSection> boolean setSectionAtomic(
+            String worldName,
+            IntPair pair,
+            LevelChunkSection[] sections,
+            LevelChunkSection expected,
+            LevelChunkSection value,
+            int layer
+    ) {
+        if (layer < 0 || layer >= sections.length) {
+            return false;
+        }
+        long[] stamp = new long[1];
+        ChunkLock[] chunkLock = new ChunkLock[1];
```

Comment:
Could these two arrays be replaced by a class? Iirc arrays have 16 byte object headers alone in current Hotspot versions, so reducing to a class would significantly cut down memory per call here (assuming it doesn't get scalar replaced, which is rather unlikely due to the complexity of the involved code).